Framework/PythonInterface/plugins/algorithms/TOFTOFMergeRuns.py

```diff
@@ -0,0 +1,190 @@
+
+    def summary(self):
+        return "Merge runs and the sample logs."
+
+    def PyInit(self):
+        """ Declare properties
+        """
+        validator = StringArrayLengthValidator()
+        validator.setLengthMin(1)
+        self.declareProperty(StringArrayProperty(name="InputWorkspaces", direction=Direction.Input, validator=validator),
+                             doc="Comma separated list of workspaces or groups of workspaces.")
+        self.declareProperty(WorkspaceProperty("OutputWorkspace", "", direction=Direction.Output),
+                             doc="Name of the workspace that will contain the merged workspaces.")
+        return
+
+    def _validate_input(self):
+        """
+        Checks for the valid input:
+            all given workspaces and/or groups must exist
+            gets names of the grouped workspaces
+        """
+        workspaces = self.getProperty("InputWorkspaces").value
+        mlzutils.ws_exist(workspaces, self.log())
+        if len(workspaces) < 1:
+            message = "List of workspaces is empty. Nothing to merge."
+            self.log().error(message)
+            raise RuntimeError(message)
+        for wsname in workspaces:
+            wks = api.AnalysisDataService.retrieve(wsname)
+            if isinstance(wks, WorkspaceGroup):
+                self.wsNames.extend(wks.getNames())
+            else:
+                self.wsNames.append(wsname)
+
+    def _can_merge(self):
+        """
+        Checks whether given workspaces can be merged
+        """
+        # mandatory properties must be identical
+        mlzutils.compare_mandatory(self.wsNames, self.mandatory_properties, self.log())
+
+        # timing (x-axis binning) must match
+        # is it possible to use WorkspaceHelpers::matchingBins from python?
+        self.timingsMatch(self.wsNames)
+
+        # Check sample logs for must have properties
+        for wsname in self.wsNames:
+            wks = api.AnalysisDataService.retrieve(wsname)
+            run = wks.getRun()
+            for prop in self.must_have_properties:
+                if not run.hasProperty(prop):
+                    message = "Error: Workspace " + wsname + " does not have property " + prop +\
+                        ". Cannot merge."
+                    self.log().error(message)
+                    raise RuntimeError(message)
+
+        # warnig if optional properties are not identical must be given
+        ws1 = api.AnalysisDataService.retrieve(self.wsNames[0])
+        run1 = ws1.getRun()
+        for wsname in self.wsNames[1:]:
+            wks = api.AnalysisDataService.retrieve(wsname)
+            run = wks.getRun()
+            mlzutils.compare_properties(run1, run, self.optional_properties, self.log(), tolerance=0.01)
+        return True
+
+    def PyExec(self):
+        """ Main execution body
+        """
+        # get list of input workspaces
+        self._validate_input()
+        workspaceCount = len(self.wsNames)
```

Comment:
Maybe it would be more clear what variables are modified by the helper methods (`_validate_input`, `_can_merge`, etc.) if instead of putting the names in `self.wsNames`, the variable was passed on / returned to/from the methods.
This is something that also happens in C++ algorithms where somehow we have developed a (probably not good) tendency to put variables as data members of the algorithm class. If it's not something that needs to be cached for a very specific reason I'd say it's better to pass it around to/from the methods.